compiler/rustc_trait_selection/src/autoderef.rs

```diff
@@ -222,19 +222,10 @@ pub fn report_autoderef_recursion_limit_error<'tcx>(tcx: TyCtxt<'tcx>, span: Spa
         Limit(0) => Limit(2),
         limit => limit * 2,
     };
-    struct_span_err!(
-        tcx.sess,
+    tcx.sess.emit_err(AutoDerefReachedRecursionLimit {
         span,
-        E0055,
-        "reached the recursion limit while auto-dereferencing `{:?}`",
-        ty
-    )
-    .span_label(span, "deref recursion limit reached")
-    .help(&format!(
-        "consider increasing the recursion limit by adding a \
-             `#![recursion_limit = \"{}\"]` attribute to your crate (`{}`)",
-        suggested_limit,
-        tcx.crate_name(LOCAL_CRATE),
-    ))
-    .emit();
+        ty: format!("{:?}", ty),
+        suggested_limit: suggested_limit.to_string(),
```

Comment:
We could implement `IntoDiagnosticArg` for this and use it in the type, rather than `to_string` here.

Reply:
I implemented it `IntoDiagnosticArg` for the types where I was calling `to_string`